frame/assets/src/functions.rs

```diff
@@ -679,35 +679,70 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 				if let Some(check_owner) = maybe_check_owner {
 					ensure!(details.owner == check_owner, Error::<T, I>::NoPermission);
 				}
+
+				ensure!(details.is_frozen, Error::<T, I>::BadWitness);
 				ensure!(details.accounts <= witness.accounts, Error::<T, I>::BadWitness);
 				ensure!(details.sufficients <= witness.sufficients, Error::<T, I>::BadWitness);
 				ensure!(details.approvals <= witness.approvals, Error::<T, I>::BadWitness);
 
-				for (who, v) in Account::<T, I>::drain_prefix(id) {
+				let mut removed_accounts = 0u32;
+				let mut removed_approvals = 0u32;
+
+				let accounts_to_delete: Vec<(T::AccountId, _)> = Account::<T, I>::iter_prefix(id)
+					.take(T::RemoveAccountsLimit::get() as usize)
+					.collect();
+
+				for (who, v) in accounts_to_delete {
+					if removed_accounts >= T::RemoveAccountsLimit::get() {
+						break
+					}
+
+					Account::<T, I>::remove(id, &who);
+
 					// We have to force this as it's destroying the entire asset class.
 					// This could mean that some accounts now have irreversibly reserved
 					// funds.
 					let _ = Self::dead_account(&who, &mut details, &v.reason, true);
 					dead_accounts.push(who);
+					removed_accounts += 1;
 				}
```

Comment:
Cool, so later I can check with [that](https://github.com/paritytech/substrate/blob/3067e381aeb0c915152db968a2ae963229f76476/frame/assets/src/functions.rs#L742-L746), if `removed_accounts.is_zero()` to know if all accounts are destroyed.